plugin/http/httpstats/transport.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpstats
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+// Transport is an http.RoundTripper that collects stats for the outgoing requests.
+type Transport struct {
+	// Base represents the underlying roundtripper that does the actual requests.
+	// If none is given, http.DefaultTransport is used.
+	//
+	// If base HTTP roundtripper implements CancelRequest,
+	// the returned round tripper will be cancelable.
+	Base http.RoundTripper
+}
+
+// RoundTrip implements http.RoundTripper, delegating to Base and recording stats for the request.
+func (t Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx, _ := tag.New(req.Context(),
+		tag.Upsert(Host, req.URL.Host),
+		tag.Upsert(Path, req.URL.Path),
+		tag.Upsert(Method, req.Method))
+	req = req.WithContext(ctx)
+	var (
+		resp *http.Response
+		err  error
+	)
+	track := &tracker{
+		start: time.Now(),
+		ctx:   ctx,
+	}
+	if req.Body == nil {
+		//TODO: handle cases where ContentLength is not set
+		track.reqSize = -1
+	} else if req.ContentLength > 0 {
+		track.reqSize = req.ContentLength
+	}
+	stats.Record(ctx, ClientRequest.M(1))
+
+	// Perform request
+	resp, err = t.base().RoundTrip(req)
+
+	if err != nil {
+		track.statusCode = "error"
+		track.end()
+	} else {
+		track.statusCode = strconv.Itoa(resp.StatusCode)
+		if resp.Body == nil {
+			track.end()
+		} else {
+			track.body, resp.Body = resp.Body, track
+		}
+	}
+
+	return resp, err
+}
+
+func (t Transport) base() http.RoundTripper {
+	if t.Base != nil {
+		return t.Base
+	}
+	return http.DefaultTransport
+}
+
+// CancelRequest cancels an in-flight request by closing its connection.
+func (t Transport) CancelRequest(req *http.Request) {
+	type canceler interface {
+		CancelRequest(*http.Request)
+	}
+	if cr, ok := t.base().(canceler); ok {
+		cr.CancelRequest(req)
+	}
+}
+
+type tracker struct {
+	respSize   int64
+	reqSize    int64
+	ctx        context.Context
+	start      time.Time
+	body       io.ReadCloser
+	statusCode string
+	endOnce    sync.Once
+}
+
+func (t *tracker) end() {
+	t.endOnce.Do(func() {
+		m := []stats.Measurement{
+			ClientLatency.M(float64(time.Since(t.start)) / float64(time.Millisecond)),
```

Comment:
Very nice!!

Reply:
Although, let's also preserve the nano seconds if we can
```go
latency := time.Since(t.start)
ms := latency / time.Millisecond
ns := latency % time.Millisecond
m := []stats.Measurement{
    ClientLatency.M(float64(ms) + (float64(ns)/1e3)),
```

Reply:
I don't see the difference between this and the existing statement - this is float division so I think it's the same unless I'm not seeing something?

Reply:
Crap, sorry for the false alarm, you are right. Thanks for the correction.